lib/smart_name/parts.rb

```diff
@@ -108,17 +99,28 @@ def pieces
         end
     end
 
-    # name parts can be accessed and manipulated like an array
-    def method_missing method, *args, &block
-      if parts.respond_to?(method)
-        self.class.new parts.send(method, *args, &block)
-      else
-        super
-      end
+    def + other
+      self.class.new(parts + other.to_name.parts)
     end
 
-    def respond_to? method, include_private=false
-      super || parts.respond_to?(method, include_private)
+    def [] *args
+      self.class.new parts[*args]
     end
+
+    # full support of array methods caused trouble with `flatten` calls
```

Comment:
If the main issue is flatten, couldn't you implement flatten correctly instead of letting it be handled by the method on the parts?
Just what would you expect smartname.flatten to do? I think it should be a no-op, return the input.